framework/src/multiapps/MultiApp.C

```diff
@@ -173,35 +171,67 @@ MultiApp::initialSetup()
 void
 MultiApp::fillPositions()
 {
+  if (isParamValid("positions") && isParamValid("positions_file"))
+    mooseError("Both 'positions' and 'positions_file' cannot be specified simultaneously in MultiApp " << name());
+
   if (isParamValid("positions"))
+  {
     _positions = getParam<std::vector<Point> >("positions");
+
+    if (_positions.size() < _input_files.size())
+      mooseError("Not enough positions for the number of input files provided in MultiApp " << name());
+  }
   else if (isParamValid("positions_file"))
   {
-    // Read the file on the root processor then broadcast it
-    if (processor_id() == 0)
+    std::vector<FileName> positions_files = getParam<std::vector<FileName> >("positions_file");
+    std::vector<FileName> input_files = getParam<std::vector<FileName> >("input_files");
+
+    if (input_files.size() != 1 && positions_files.size() != input_files.size())
+      mooseError("Number of input_files for MultiApp " << name() << " must either be only one or match the number of positions_file files");
+
+    // Clear out the _input_files because we're going to rebuild it
+    if (input_files.size() != 1)
+      _input_files.clear();
+
+    for (unsigned int p_file_it = 0; p_file_it < positions_files.size(); p_file_it++)
```

Comment:
Well - I needed a counter at the time and didn't want to name it `k`.  I could change it now to an actual iterator.  The current code doesn't rely on it being a counter.